README.md

```diff
@@ -1,34 +1,332 @@
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+Main components are listed below.
+- A `Resource` trait representing for all kinds of resources.
+```rust
+pub trait Resource {}
+```
+
+- A `ResourceSize` trait representing for all kinds of resource size.
+```rust
+pub trait ResourceSize {}
+```
+
+- A `ResourceAllocator` trait representing for all kinds of resources allocation, freeing
+and updating.
+```rust
+pub trait ResourceAllocator<T: Resource, S: ResourceSize> {
+    /// Allocate some resource with given `resource` and `size`.
+    ///
+    /// # Arguments
+    ///
+    /// * `resource`: resource to be allocate.
+    /// * `size`: resource size of allocation request.
+    fn allocate(
+        &mut self,
+        resource: Option<T>,
+        size: S,
+   ) -> Result<T, Error>;
+
+    /// Free resource specified by given `resource` and `size`.
+    ///
+    /// # Arguments
+    ///
+    /// * `resource`: resource to be free.
+    /// * `size`: resource size of free request.
+    fn free(&mut self, resource: T, size: S);
+
+    /// Update resource, freeing the old one and allocating a new range.
+    ///
+    /// This is mostly used when guest allocated some resources or reprogrammed,
+    /// so the new allocated resources should be marked as used, and the preallocated
+    /// one should be freed (if has). 
+    ///
+    /// # Arguments
+    ///
+    /// * `old_resource`: old resource to be free or no such preallocated resources.
+    /// * `old_size`: old resource size of free request or no such preallocated resources.
+    /// * `resource`: resource to be free.
+    /// * `size`: resource size of free request.
+    fn update(&mut self, old_resource: Option<T>, old_size: Option<S>, new_resource: T, new_size: S);
+}
+```
+
+Different kinds of resources and its allocators should implement the three traits. Take
+address space resource for example.
+
+```rust
+impl Resource for GuestAddress {}
+
+// This should be implemented for u64.
+// Otherwise implementing for GuestUsize (aka. <GuestAddress as AddressValue>::V)
+// would impact other implementation such as instance id since compiler
+// doesn't know real type of V.
+impl ResourceSize for u64 {}
+
+pub struct AddressAllocator {
+    base: GuestAddress,
+    end: GuestAddress,
+    alignment: GuestUsize,
+    ranges: BTreeMap<GuestAddress, GuestUsize>,
+}
+
+impl ResourceAllocator<GuestAddress, GuestUsize> for AddressAllocator {
+    /// Allocates a range of addresses from the managed region.
+    fn allocate(
+        &mut self,
+        address: Option<GuestAddress>,
+        size: GuestUsize,
+    ) -> Result<GuestAddress> {...}
+
+    /// Free an already allocated address range.
+    fn free(&mut self, address: GuestAddress, size: GuestUsize) {...}
+
+    /// Update an address range with a new one.
+    fn update(&mut self, old_resource: Option<T>, old_size: Option<S>, new_resource: T, new_size: S) {...}
```

Comment:
With the parameters this function currently has it looks more like a `remove_and_replace` than an update.

The `update` function does not make sense in the context of `IdAllocator` I think. Also in the context of other allocators. Another reason for not having a `ResourceAllocator` as a trait because not all resources can follow the same interface.